### PR TITLE
Ensure we fetch a valid deployed production hash

### PR DESCRIPTION
**Pull request by jharrington22**
This PR does two things to discover and ensure we fetch a valid HASH.

1. We determine if the SAAS file is applying a resource type of `SelectorSyncSet` if it is we'll assume the saas file is for a customer cluster and not Hive. (Resources cannot be sync'd to a customer cluster without a SelectorSyncSet (I should validate that and SyncSet).
2. Once we know the resource type we can make an assumption on the namespace path reference in resourceTemplates.

Before adding validation (Non hive cluster):

```
Image quay.io/jharrington22/cloud-ingress-operator-registry:production-1a3ab03 was deleted from the repository.
Proceeding as if it never existed.
make[1]: Entering directory '/home/jamesh/.gvm/pkgsets/go1.13.6/global/src/github.com/openshift/cloud-ingress-operator'
Cloning into 'saas-cloud-ingress-operator-bundle'...
remote: Enumerating objects: 155, done.
remote: Counting objects: 100% (155/155), done.
remote: Compressing objects: 100% (50/50), done.
remote: Total 1348 (delta 97), reused 138 (delta 81), pack-reused 1193
Receiving objects: 100% (1348/1348), 130.62 KiB | 228.00 KiB/s, done.
Resolving deltas: 100% (891/891), done.
SAAS file is NOT applied to Hive, MANAGED_RESOURCE_TYPE=SelectorSyncSet
Current deployed production HASH: 
Error discovering current production deployed HASH
make[1]: *** [boilerplate/openshift/golang-osd-operator/csv-generate/csv-generate.mk:42: production-common-csv-build] Error 1
make[1]: Leaving directory '/home/jamesh/.gvm/pkgsets/go1.13.6/global/src/github.com/openshift/cloud-ingress-operator'
make: *** [boilerplate/openshift/golang-osd-operator/standard.mk:211: build-push] Error 2

```

After (Non hive cluster)

```
Image quay.io/jharrington22/cloud-ingress-operator-registry:staging-1a3ab03 exists with digest sha256:64b5c9d1090c21f67419a282ea99971d62ab23ea3e53c05e2c4bae622d09e0ec.                                                                                       
Catalog image quay.io/jharrington22/cloud-ingress-operator-registry:staging-1a3ab03 already                                                                                                                                                                   
exists. Assuming this means the saas bundle work has also been done                                                                                                                                                                                           
properly. Nothing to do!                                                                                                                                                                                                                                      
Image quay.io/jharrington22/cloud-ingress-operator-registry:production-1a3ab03 was deleted from the repository.                                                                                                                                               
Proceeding as if it never existed.                                                                                                                                                                                                                            
make[1]: Entering directory '/home/jamesh/.gvm/pkgsets/go1.13.6/global/src/github.com/openshift/cloud-ingress-operator'                                                                                                                                       
Cloning into 'saas-cloud-ingress-operator-bundle'...                                                                                                                                                                                                          
remote: Enumerating objects: 155, done.                                                                                                                                                                                                                       
remote: Counting objects: 100% (155/155), done.                                                                                                                                                                                                               
remote: Compressing objects: 100% (50/50), done.                                                                                                                                                                                                              
remote: Total 1348 (delta 97), reused 138 (delta 81), pack-reused 1193                                                                                                                                                                                        
Receiving objects: 100% (1348/1348), 130.62 KiB | 165.00 KiB/s, done.                                                                                                                                                                                         
Resolving deltas: 100% (891/891), done.                                                                                                                                                                                                                       
SAAS file is NOT applied to Hive, MANAGED_RESOURCE_TYPE=SelectorSyncSet                                                                                                                                                                                       
Current deployed production HASH: 10af8dab0515ed5762aa283c83eb4080b0238fcd                                                                                                                                                                                    
Generating CSV for version: 0.1.344-1a3ab03    
```

We can go further to validate this by running the built catalog images and comparing the versions returned by the registry API and what we expect according to the `saas-${operator_name}-bundle` repo. This is how I realized the issue in the first place.

### CODE REVIEW

**Review by 2uasimojo — COMMENTED** [review]
Doesn’t it depend whether we’re building a hive operator vs a cluster operation which way this needs to go?

**Comment by maorfr** [comment]
https://github.com/openshift/aws-account-operator/pull/493

**Comment by 2uasimojo** [comment]
I guess the concern with switching to graphql is that it requires yet another auth token for local testing. And iirc that one isn’t trivial to get — requires a MR in app-interface and pgp and...

**Comment by maorfr** [comment]
And that's it.

**Comment by jharrington22** [comment]
> I guess the concern with switching to graphql is that it requires yet another auth token for local testing. And iirc that one isn’t trivial to get — requires a MR in app-interface and pgp and...

Exactly. This works as is and gives us _some_ safety. ++ to switching to graphql but WHEN we have boilerplate we can do it all at once.

**Comment by maorfr** [comment]
App-interface being in gitlab is an implementation detail you shouldn't rely on. It also restricts us from moving it, which we may consider in the future. It also forces us to run these jobs behind the VPN so the clone will work.

Worse, in case we decide to change the schema structure, we suddenly have a fire on our hands as all your jobs will start to fail.
The (graphql) API is what you SHOULD be using.

The concern about the MR - it takes 2 minutes. And you should probably have your gpg keys in app-interface as part of your onboarding process. It has many other usages that you may need to use in the future.

I can go on forever 😉

**Comment by jharrington22** [comment]
> App-interface being in gitlab is an implementation detail you shouldn't rely on. It also restricts us from moving it, which we may consider in the future. It also forces us to run these jobs behind the VPN so the clone will work.
> 
> Worse, in case we decide to change the schema structure, we suddenly have a fire on our hands as all your jobs will start to fail.
> The (graphql) API is what you SHOULD be using.
> 
> The concern about the MR - it takes 2 minutes. And you should probably have your gpg keys in app-interface as part of your onboarding process. It has many other usages that you may need to use in the future.
> 
> I can go on forever

1000% agree @maorfr are the ENV credentials that we need already exist? Mostly I'm trying to change as few pieces as possible while we migrate to boilerplate everywhere. I'm pretty sure that a schema change is the cause of all our OLM bugs in prod (will link later) so I want to get here as soon as we can.

**Comment by maorfr** [comment]
Everything is ready and a working example is the PR I mentioned above.

**Comment by maorfr** [comment]
For the sake of splitting efforts, this change is LGTM, and a switch to graphql is not urgent at all.

**Comment by jharrington22** [comment]
@maorfr if everything is there I'm all in :slightly_smiling_face: I thought we had to make a couple of MR's get what we need available to the job.

**Comment by 2uasimojo** [comment]
> @maorfr if everything is there I'm all in  I thought we had to make a couple of MR's get what we need available to the job.

The *job* already has what it needs. It's local testing that would be painful. So I'm good with it as long as the PR that cuts over to graphql also updates the [testing doc](https://github.com/openshift/boilerplate/blob/f2167d69d378ebb468a49e54f2cae068ab10a383/boilerplate/openshift/golang-osd-operator/app-sre.md) to describe how to set up those creds. Most of that is covered in [these instructions](https://gitlab.cee.redhat.com/service/app-interface#querying-the-app-interface), but it would be useful to also a) point to an [example MR](https://gitlab.cee.redhat.com/service/app-interface/-/merge_requests/10873) and b) go into a little more detail about how to configure the creds in the environment (I recall struggling with that a little bit).

File: boilerplate/openshift/golang-osd-operator/csv-generate/csv-generate.sh
Comment (2uasimojo) [inline boilerplate/openshift/golang-osd-operator/csv-generate/csv-generate.sh:88]: Did we ascertain that this will work for all the operators? Based on what's in [that directory](https://gitlab.cee.redhat.com/service/app-interface/-/tree/master/data/services/osd-operators/namespaces/hivep01ue1), I'm guessing not -- we would need to condition cluster- vs hive-scoped operators somehow.

**Comment by jharrington22** [comment]
@2uasimojo lets merge as is, I'll create a card to follow up and change it to graphql once we have all the pieces sorted.

File: boilerplate/openshift/golang-osd-operator/csv-generate/csv-generate.sh
Comment (jharrington22) [reply boilerplate/openshift/golang-osd-operator/csv-generate/csv-generate.sh:88]: let me validate.

**Comment by 2uasimojo** [comment]
> @2uasimojo lets merge as is, I'll create a card to follow up and change it to graphql once we have all the pieces sorted.

I think we crossed in the mail -- please see above [comment](https://github.com/openshift/boilerplate/pull/139#discussion_r613386684)

You would also need to fix CI :)

**Comment by jharrington22** [comment]
Yeah @2uasimojo I need to update this concludes our deploy issues, its not the same. 

/hold

**Comment by jharrington22** [comment]
Testing this solution

File: boilerplate/openshift/golang-osd-operator/csv-generate/csv-generate.sh
Comment (2uasimojo) [inline boilerplate/openshift/golang-osd-operator/csv-generate/csv-generate.sh:109]:
```suggestion
        HIVE=false
```
?

File: boilerplate/openshift/golang-osd-operator/csv-generate/csv-generate.sh
Comment (2uasimojo) [inline boilerplate/openshift/golang-osd-operator/csv-generate/csv-generate.sh:117]: Why not just set a variable?
```bash
if $HIVE; then
    filename=${operator_name}.yml
else
    filename=cluster-scope.yml
fi
...
docker run ... "...\$ref==/services/osd-operators/namespaces/hivep01ue1/${filename}).ref"
```

**Review by 2uasimojo — CHANGES_REQUESTED** [review]
This looks workable except for the mistake noted inline

but

Perhaps it would be easier to require the consumer to set a Makefile variable indicating whether they're hive or not. The less we can rely on specific yaml formats (or even graphql queries) the better.

**Review by 2uasimojo — APPROVED** [review]

File: boilerplate/openshift/golang-osd-operator/csv-generate/csv-generate.sh
Comment (jharrington22) [reply boilerplate/openshift/golang-osd-operator/csv-generate/csv-generate.sh:117]: Good idea.

File: boilerplate/openshift/golang-osd-operator/csv-generate/csv-generate.sh
Comment (jharrington22) [reply boilerplate/openshift/golang-osd-operator/csv-generate/csv-generate.sh]: Fixed.

File: boilerplate/openshift/golang-osd-operator/csv-generate/csv-generate.sh
Comment (jharrington22) [reply boilerplate/openshift/golang-osd-operator/csv-generate/csv-generate.sh]: redundant now.

Comment (jharrington22) [reply boilerplate/openshift/golang-osd-operator/csv-generate/csv-generate.sh]: > This looks workable except for the mistake noted inline
> 
> but
> 
> Perhaps it would be easier to require the consumer to set a Makefile variable indicating whether they're hive or not. The less we can rely on specific yaml formats (or even graphql queries) the better.

I don't disagree with this. We have protection in that if we guess wrong the hash should be blank and we'll error. In the interest of time I think lets move forward with this (since its only one merge and we don't have to touch the other operators) and fix this later. https://issues.redhat.com/browse/OSD-6995

File: boilerplate/openshift/golang-osd-operator/csv-generate/csv-generate.sh
Comment (2uasimojo) [inline boilerplate/openshift/golang-osd-operator/csv-generate/csv-generate.sh:98]: `docker` will need to be conditioned. Something like [this](https://github.com/openshift/boilerplate/pull/142/files#diff-424ddf34b68cef0633f4520f756a919d70a24880b5ab931d3cd85d17f8659b5fR119-R126) :P

Comment (jharrington22) [reply boilerplate/openshift/golang-osd-operator/csv-generate/csv-generate.sh]: Fixed.

**Comment by 2uasimojo** [comment]
This should be rebased now that #142 has merged.

The assignment of `CONTAINER_ENGINE` should move up to where you've got it.

And every invocation that might be containerized will need to be conditioned on that variable being set or not.

**Review by 2uasimojo — APPROVED** [review]
Nice

/lgtm

**Comment by 2uasimojo** [comment]
Okay, so CI was failing because the gitlab query was coming up empty. We thought it was because we were using a dummy project that didn't have a SaaS file in app-interface. So I first tried adding an override to point the `SAAS_FILE_URL` to a known project's SaaS file.

But it still didn't work -- the curl command was producing no output and exiting with RC=60 ("Peer certificate cannot be authenticated with known CA certificates.")

Turns out the CI environment doesn't have the right certificates to curl gitlab (or maybe anything?) over https. So I *also* added a granny switch to curl with `--insecure` and enabled it in the test case.

**Comment by 2uasimojo** [comment]
I'm going to stamp this so it lands if CI passes.

/lgtm

This will unblock us and allow us to propose bp updates in consumer projects.

**Comment by 2uasimojo** [comment]
Okay, at this point I'm guessing it's just that the CI env doesn't have access to gitlab at all (like, it's not inside the VPN?). So I'm going back to disabling that whole chunk.

**Comment by 2uasimojo** [comment]
And

/lgtm

again, for the [same reasons](https://github.com/openshift/boilerplate/pull/139#issuecomment-820074613).

**Pull request merged**
